Responsive page (document) titles with dynamic data! Fixes issue #333

ktuite opened the pull request:
Work in progress, just putting it online for feedback! Related to issue #333 

<img width="1110" alt="Screen Shot 2021-04-01 at 5 02 49 PM" src="https://user-images.githubusercontent.com/76205/113365843-4ddaaf00-930c-11eb-81f3-ac5211d87654.png">
<img width="1110" alt="Screen Shot 2021-04-01 at 5 01 31 PM" src="https://user-images.githubusercontent.com/76205/113365848-4fa47280-930c-11eb-9a76-eaf882596886.png">


Explaining some of my rationale
* Adding titles in `routes.js` feels natural (having them all in one place vs. some other document mapping routes to names elsewhere)
    * But then I wrote a whole thing in `locales/en.json5` to come up with names
    * Currently... my new translations are broken and just coming through as "project.Overview | My Project Name | ODK Central"
* I made ‘title’ a function in meta
    * Mainly because I saw one example online of passing the route into it
    * I considered passing the vuex store in, but then imported it directly.
* Adding afterEach functionality to the router to set document.title is what many other people do (sometimes people use beforeEach instead)
* Sometimes we already have all the information we need
    * Static title fully stored in the route
    * Project or form name from vuex
* Sometimes we don’t… the router has a way of watching vuex store for changes for some other purposes (validateData)   
    * So in a similar way, I use the route’s meta field to watch for changes to a particular store key
* If the title is supposed to use delayed information… don’t think too hard about an alternate title (e.g. "Project Overview" instead of "Project Name")
* Page title gets updated when locale changes
* Parts of titles are generated as an array and joined with `I`, e.g. `"Form Access | My Project Name | ODK Central"` (all have ODK Central at the end)
    * Can change this general formatting in `src/util/router.js`
 * Nothing has a test yet
 
<img width="1110" alt="Screen Shot 2021-04-01 at 5 03 52 PM" src="https://user-images.githubusercontent.com/76205/113365856-57fcad80-930c-11eb-8b3f-f29fa3c6d244.png">


If there's anything weird about my javascript that I can improve, please let me know!

If people have thoughts about how to be more consistent with titles (either how they look (patterns, breadcrumb kinda things, how many nesting layers), or where they live in the code, especially with respect to translations) let's discuss!

matthew-white commented:
Very excited to see this! This definitely looks like it's going in the right direction.

> Adding titles in `routes.js` feels natural

That sounds right to me!

> But then I wrote a whole thing in `locales/en.json5` to come up with names

That makes sense to me to add these messages to `en.json5`. Since `routes.js` is outside a component, you can't use a single file component `i18n` custom block, which means the messages have to live in `en.json5`.

> Currently... my new translations are broken and just coming through as "project.Overview | My Project Name | ODK Central"

I mentioned this over Slack, but I think the message path just needs the `title.` prefix.

> Adding afterEach functionality to the router to set document.title is what many other people do (sometimes people use beforeEach instead)

`afterEach()` makes sense to me, since after `beforeEach()` is called, there may still be a redirect, or the navigation may be canceled. And in Frontend, navigation is synchronous after the initial navigation, so `beforeEach()` and `afterEach()` will be called in the same tick anyway.

> * I made ‘title’ a function in meta
>     * Mainly because I saw one example online of passing the route into it
>     * I considered passing the vuex store in, but then imported it directly.

If you pass the store to `updateDocumentTitle()`, it could pass `store.state.request.data` to the `title` function, which could destructure it:

```
title: ({ project }) => [
  i18n.t('title.project.overview'),
  project != null ? project : project.name
]
```

Just a thought! Also, at some point soon, we'll be able to use `?.` rather than a ternary.

> * Sometimes we don’t… the router has a way of watching vuex store for changes for some other purposes (validateData)
>     * So in a similar way, I use the route’s meta field to watch for changes to a particular store key

Perfect! I think it makes a lot to sense to use that pattern here.

> If the title is supposed to use delayed information… don’t think too hard about an alternate title (e.g. "Project Overview" instead of "Project Name")

I think right now, if you're on the project overview page, but haven't received the project yet, the `title` function will return `['Overview', null]`, and `updateDocumentTitle()` will set the title to `Overview |  | ODK Central`. Not sure how best to address that case... One option is that `updateDocumentTitle()` could filter out `null` before joining. Alternatively, `updateDocumentTitle()` could check whether data exists for the request key specified to `delayedData`, and if not, skip calling the `title` function and simply set the title to `ODK Central`. Or maybe there's a third option. 🤔

Speaking of `delayedData`, it'd be amazing if `delayedData` and `title` could be combined somehow, since `delayedData` only relates to `title` and not other things waiting for data (like `validateData`). Maybe `title` could be an array that combines the request key and the `title` function?

```
title: [
  'project',
  () => [i18n.t('title.project.overview'), getResourceNameFromStore('project', store)]
]
```

By the way, not all resources have a `name` property. For example, a user has a `displayName` property. Also, I believe the `name` property of a form is optional. `AuditRow` contains some logic that I think is related (though I'm not sure it could be reused):

https://github.com/getodk/central-frontend/blob/7791afbd75726357599efc34c6640586c554bddb/src/components/audit/row.vue#L58-L81

> Page title gets updated when locale changes

Awesome! 👏

> If people have thoughts about how to be more consistent with titles (either how they look (patterns, breadcrumb kinda things, how many nesting layers)

I think the current pattern you're using is great. You show a single resource name even for nested resources, which I think makes sense and helps ensure the title isn't too long. @lognaturel, it'd be great to get your thoughts as well, since you made a [related comment](https://github.com/getodk/central-frontend/issues/333#issuecomment-654466975) on the issue.

lognaturel commented:
This is great! 🎉

First, the world's tiniest thing that we can ignore -- I noticed while using Spotify that · seems to be the more modern equivalent to |. You'll see it on Github as well. I like | but I have to say that it's giving me 90s vibes. 😄

I still agree with my prior comment that for project-level things the static part is most helpful (App Users, Roles, etc). It's probably simplest to make that pattern consistent but I do feel a bit sad that those of us who run with a lot of tabs won't see the dynamic part at a glance.

A common state I end up in is having two forms open on their overview page as I'm updating drafts in parallel (e.g. baseline/followup). Would it be crazy to have a special case for both project and form overviews that drop the "Overview |" prefix and just show the form/project name? Similarly, once the submission page is up, it'd be great for that to show the instance name (won't be so fun if it's a uuid but still more helpful than leading with Overview, I think).

matthew-white commented on src/router.js:
```diff
@@ -167,6 +167,11 @@ of the `key` attribute.)
     }
+
+    const key = to.meta.delayedData;
+    unwatch.push(store.watch((state) => state.request.data[key], () => {
+      updateDocumentTitle(to);
+    }));
```
There's one other case I just thought of, which is that data can be cleared and re-fetched while the user is on a page. It's pretty uncommon, and the only example I can think of that would be relevant here is enabling encryption: if you enable encryption on a project, the `project` data in the store is cleared, and a new request for the project is sent. Clearing the project will trigger this watcher on the store, which will reset the title. That might actually be the behavior we want — mostly I just wanted to put the scenario on your radar.

ktuite replied:
Got it! Updating the title in this case of encryption changing, even if the project name didn't change, seems ok.

ktuite commented:
> A common state I end up in is having two forms open on their overview page as I'm updating drafts in parallel (e.g. baseline/followup). Would it be crazy to have a special case for both project and form overviews that drop the "Overview |" prefix and just show the form/project name? Similarly, once the submission page is up, it'd be great for that to show the instance name (won't be so fun if it's a uuid but still more helpful than leading with Overview, I think).

Check it out @lognaturel, I've removed the "Overview" part of these root project and form pages so it will say:

    My Project | ODK Central
    Project Roles | My Project | ODK Central
    
    My Form | ODK Central
    Submissions | My Form | ODK Central

It's very easy to add as many or as few of these chunks between the pipes as you want!
I might have gone overboard on the system management tab names: (should they say something else?)

    Server Audit Logs | System | ODK Central
    Backups | System | ODK Central

lognaturel commented:
> I've removed the "Overview" part of these root project and form pages

That makes a lot of sense to me! I didn't look at the implementation -- it's not crazy to have different rules for each page type? Does that mean each page type defines its own title rules? I don't have enough context to have strong feelings about this but just flagging that there are options about where redundancy is captured.

> I might have gone overboard on the system management tab names: (should they say something else?)

This I don't have any strong feelings on either. The important thing to me is that the differentiating part is on the left. "System" doesn't add much to me but I also don't feel like it matters. Seems reasonable to leave in case we end up having project-only subsets of the same data at some point, for example.

ktuite commented:
> That makes a lot of sense to me! I didn't look at the implementation -- it's not crazy to have different rules for each page type? Does that mean each page type defines its own title rules? I don't have enough context to have strong feelings about this but just flagging that there are options about where redundancy is captured.

There's not really a more standardized or centralized place to pull page titles from, so yes, there's basically a custom rule for each route. It's in the `src/routes.js` file, which is where other people put their static page titles. We've just got functions instead (that @matthew-white and I have tried to make as concise as possible) that account for translations and fields that may or may not be present yet on various objects.

Each route has its own function for building an array of title "parts", which can be as simple as ["Log In"] (but with i18n also thrown in there) or more complicated like ["App Users", project.name]. This array is then combined with the standard ["ODK Central"] and everything is joined with the separator.

matthew-white reviewed (commented):
This approach is looking good! I think this approach does a great job of providing flexibility while minimizing boilerplate. I've added specific questions/comments below, but it feels like this is getting pretty close.

At some point soon, it'd also be great to add tests of the new functionality. We have some unit tests of the functions in `src/util/router.js` and then a lot of tests of the router itself that are more integration tests. You'll want to write tests of the router, but I'm less sure that unit tests are needed here.

The tests will likely use `load()`, so if you haven't already, you should read the comments at the top of `test/util/http.js`, which I updated pretty recently. I think things there are reasonably well documented at this point, but would definitely love to hear about any gaps in documentation.

You're welcome to add a test for each route. However, since there's so much similarity among the routes, and it's a low-risk feature, it's probably also OK to test a smaller number of routes that are representative.

matthew-white commented:
I mentioned this a little earlier today, but I just merged a couple of PRs that I think are relevant to this PR:

- #429 adds a new route (the submission detail page).
- #430 defines a utility function that takes a submission and returns either its instance name or its instance ID.

ktuite commented on src/router.js:
```diff
@@ -167,6 +167,13 @@ of the `key` attribute.)
       }));
     }
+
+    if (to.meta.title.key != null) {
```
This is what I saw in various examples online (https://dev.to/webhookrelay/a-simple-way-to-keep-your-vue-page-title-in-sync-with-the-router-ec0) but you're right, it doesn't seem to be documented.

Deep in the vue router source code, I see it pulling `meta` from the Route object RouteRecord: https://github.com/vuejs/vue-router/blob/dev/src/util/route.js#L23

matthew-white replied:
Very interesting! In general, I feel like it's a little unclear where the `Route` ends and the `RouteRecord` begins. For example, it looks from the source like the `Route` is also taking the `name` from the `RouteRecord` (though that seems better documented). Vue Router has previously changed undocumented behavior that we used in Frontend, so I do think it'd be preferable to use `to.matched`. I see that you're destructuring `meta` from `last(to.matched)` above, so I think the only thing left is to use `meta` here and in the line below.

ktuite commented on src/util/router.js:
```diff
@@ -81,3 +81,19 @@ export const canRoute = (to, from, store) => {
+    resource object (e.g. project or form).
+*/
+export const updateDocumentTitle = (to, store) => {
+  const titlePath = (to.meta.title.parts != null)
```
If someone defines this `meta.title` but doesn't add `parts()` to it, is there a short way to get `normalizeMeta` to check for that case and add it?

Or at this point.... maybe we can assume if the developer adds `title` without `parts` that's on them (and the error message in this case is informative talking about `parts` not existing).

ktuite replied:
I took out this check and am assuming parts will be there if title is there.

ktuite commented on src/routes.js:
```diff
@@ -440,13 +547,17 @@ const routes = [
           currentUser.can(['user.read', 'user.update'])
-      }
+      },
+      title: { parts: () => [i18n.t('title.editUser')] }
```
Yes! I just made it say "Display Name | ODK Central" rather than have "Edit" in there because that's what the page header does.

ktuite commented on src/locales/en.json5:
```diff
@@ -1,10 +1,42 @@
 // tab: This is the text of a navigation tab.
 
 {
+  "title": {
```
Ok! I've now got

    // title: This is text of the document/page title shown in the browser tab and in browser history.

ktuite commented on src/routes.js:
```diff
@@ -297,6 +350,10 @@ const routes = [
+          },
+          title: {
+            key: 'form', // store key for title data that may be delayed
+            parts: ({ form }) => [form != null ? form.name : null]
```
Made this change!

ktuite commented on src/routes.js:
```diff
@@ -212,6 +233,10 @@ const routes = [
             project: (project) => project.permits('form.list')
+          },
+          title: {
+            key: 'project', // store key for title data that may be delayed
```
I've added some comments in `src/routes.js` and `CONTRIBUTING.md`. Please take a look!

matthew-white replied:
Looks great! 👏

matthew-white commented on src/routes.js:
```diff
@@ -299,6 +392,10 @@ const routes = [
+          },
+          title: {
+            key: 'form',
+            parts: ({ form }) => [form != null ? (form.name || form.xmlFormId) : null]
```
This fallback comes up often enough that there's actually a method for it: `form.nameOrId()`.

matthew-white commented on src/locales/en.json5:
```diff
@@ -1,10 +1,43 @@
 // field: This is the text of a form field.
 // header: This is the text of a table column header.
-
+// title: This is text of the document/page title shown in the browser tab and in browser history.
```
This comment looks great — I think it'll be helpful to translators. You can actually add it right above `"title": {` below. Doing so will make sure the comment is visible in Transifex for each message in the `title` object. Defining the comment at the top of `en.json5` would apply it even more broadly: not just to this `title` message, but to any message with a `title` key, including in components. (And actually, if you look at the autogenerated `strings_en.json` below, you can see that a couple of unrelated messages received the comment.)

matthew-white reviewed (commented):
I know you're still working on things, but I thought I'd leave a few small comments. This is coming together really nicely! ✨

ktuite commented on src/locales/en.json5:
```diff
@@ -1,10 +1,42 @@
 // tab: This is the text of a navigation tab.
 
 {
+  "title": {
```
This sounds good! Do you have suggestions on how to approach this?

1. Should everything just go into `src/locales/en.json5`?
2. What about the previously-translated page titles for these components? How can I preserve that translation work?

matthew-white replied:
I do think that each title (or title part) will need a corresponding message in `en.json5`, so if a title is defined in a component, it should be moved to `en.json5`. Unfortunately, that will result in some work for translators, because the new messages in `en.json5` will appear as new strings to translate. For that reason, it's generally a good idea to avoid moving a message unnecessarily. That said, Transifex makes the process much easier, because Transifex will remember the previous translation and will offer it as a suggestion. So if there's a reason to move a message, it's also OK to do so.

One thing to look out for is that some of the messages used in components actually already live in `en.json5`. For example, the text "App Users" for the App Users tab lives in `en.json5` at `resource.appUsers`. I think the easiest thing to do would be to keep those messages where they are, then use them in `routes.js`. How does that sound? You could also considering adding some messages to the `resource` object rather than the `title` object. Like maybe "Projects" could go under `resource.projects`?

ktuite commented:
I'm feeling good about the state of this PR!
* There are tests for every route (as I was moving translation strings around, I temporarily broke route titles, so decided they should all be tested)
* There is a test for changing the location through the dropdown and having the title update right then
* Translation strings are more centralized in `src/locales/en.json5`:
    * I used existing strings from `resource.` or `action.` where I could
    * I moved certain strings out of components, but they are named after their components (e.g. `formHead` for `src/components/form/head.vue`)
    * Everything that was a "tab" is still a "tab" and has the tab transifex comment
    * I added a few other transifex comments
* User display name is shown when editing that user
* Submission UUID is shown when viewing that submission

matthew-white reviewed (commented):
This is looking good! :+1: Titles are using resource names and IDs. Messages have been moved from components and centralized in `en.json5`. The tests look excellent and seem to cover all the cases.

Going forward, it feels clear where to define the title of a new nested route. If the route is a child of an existing route, then there will be a new `tab` message under the parent route. If it is a new route entirely, then there will be a new top-level object in `en.json5` with a `tab` property and an optional `title` property for the parent route itself.

However, it feels less clear where to define the title of a new top-level route. Right now these include `AccountLogin`, `AccountResetPassword`, `AccountClaim`, `AccountEdit`, `Download`, and `NotFound`. It looks like you're using `action` messages for the first five, then defining the message for `NotFound` in `util`. However, I think I wouldn't use the `action` messages for this, since those are for buttons and links and aren't title case. For example, I think the title for `AccountClaim` should be "Claim Account" in title case in order to be consistent with other page titles and to match the previous string in Transifex.

Actually, it's not uncommon in Frontend to have two versions of a message, one in title case and one not. ("Reset Password" vs. "Reset password" is an example of that.) We considered writing a Vue I18n [modifier](https://kazupon.github.io/vue-i18n/guide/messages.html#formatting-linked-locale-messages) to convert a message to title case, because that would reduce the number of strings in Transifex. (And apparently many languages don't have a specialized title case.) We decided against it at the time, because that logic would be locale-specific, and it's pretty easy in Transifex to reuse a translation.

matthew-white commented on src/locales/en.json5:
```diff
@@ -1,10 +1,42 @@
 {
+  // This is the text shown as the title of a page including in the browser tab.
+  "title": {
+    "resetPassword": "Reset Password",
```
It'd be great to remove the `title` message from `AccountResetPassword` as well.

matthew-white commented on src/locales/en.json5:
```diff
@@ -1,10 +1,42 @@
+  // This is the text shown as the title of a page including in the browser tab.
+  "title": {
+    "resetPassword": "Reset Password",
+    "claimAccount": "Claim Account",
```
We discussed this a little over Slack, but right now the title of the panel in `AccountClaim` is "Set Password". What do you think about reusing that text for the document title?

ktuite replied:
I've changed the title of this AccountClaim to also say "Set Password"... and to have that text come from en.json5

matthew-white commented on src/routes.js:
```diff
@@ -500,14 +665,21 @@ const routes = [
-    key: () => '/dl'
+    key: () => '/dl',
+    meta: {
+      title: { parts: () => [i18n.t('title.download')] }
```
I actually don't think I see a `title.download` message.

ktuite replied:
Whoops, added it back! This is one route i'm not sure how to test.

matthew-white commented on src/routes.js:
```diff
@@ -181,22 +223,42 @@ const routes = [
+    meta: {
+      requireLogin: false,
+      requireAnonymity: true,
+      title: { parts: () => [i18n.t('action.logIn')] }
```
For posterity, we decided over Slack to use `action.logIn` rather than defining a separate `title.logIn` message. That's because "Log in" is the same whether or not it's in title case, and title case doesn't seem to be used in our other locales (and is maybe just uncommon in general?).

matthew-white reviewed (commented):
Just small i18n questions!

matthew-white reviewed (approved):
Nice work, @ktuite! 🎉
